src/noise/rate_limiter.rs

```diff
@@ -0,0 +1,216 @@
+            secret_key: make_array(X25519SecretKey::new().as_bytes()),
+            start_time: Instant::now(),
+            nonce_ctr0: AtomicU64::new(0),
+            nonce_ctr1: AtomicU64::new(0),
+            mac1_key: Blake2s::new_hash()
+                .hash(LABEL_MAC1)
+                .hash(public_key.as_bytes())
+                .finalize(),
+            mac2_key: Blake2s::new_hash()
+                .hash(LABEL_COOKIE)
+                .hash(public_key.as_bytes())
+                .finalize(),
+            limit,
+            count: AtomicU64::new(0),
+            last_reset: spin::Mutex::new(Instant::now()),
+        }
+    }
+
+    /// Reset packet count (ideally should be called with a period of 1 second)
+    pub fn reset_count(&self) {
+        // The rate limiter is not very accurate, but at the scale we care about it doesn't matter much
+        let current_time = Instant::now();
+        let mut last_reset_time = self.last_reset.lock();
+        if current_time.duration_since(*last_reset_time).as_secs() >= RESET_PERIOD {
+            self.count.store(0, Ordering::SeqCst);
+            *last_reset_time = current_time;
+        }
+    }
+
+    // Compute the correct cookie value based on the current secret value and the source IP
+    // adjustement can be used to compute past values if desired (for example using adjustment = 1 will
+    // return the cookie using the previous secret, and using adjustment = 2 will return the one before
+    // that).
+    fn current_cookie(&self, addr: IpAddr, adjustment: u64) -> Cookie {
+        let mut cookie = [0u8; COOKIE_SIZE];
+        let mut secret = [0u8; 16];
+        let mut addr_bytes = [0u8; 16];
+
+        match addr {
+            IpAddr::V4(a) => addr_bytes[..4].copy_from_slice(&a.octets()[..]),
+            IpAddr::V6(a) => addr_bytes[..].copy_from_slice(&a.octets()[..]),
+        }
+
+        // The current cookie for a given IP is the MAC(responder.changing_secret_every_two_minutes, initiator.ip_address)
+        // First we derive the secret from the current time, the value of cur_counter would change with time.
+        let cur_counter = Instant::now().duration_since(self.start_time).as_secs() / COOKIE_REFRESH;
+        assert_eq!(
+            COOKIE_SIZE,
+            ChaCha20Poly1305::new_aead(&self.secret_key[..]).seal_wg(
+                cur_counter - adjustment,
+                &addr_bytes[..],
+                &[],
+                &mut secret[..],
+            )
+        );
+
+        // Next we derive the cookie
+        cookie[..].copy_from_slice(
+            &Blake2s::new_mac(&secret[..])
+                .hash(&addr_bytes[..])
+                .finalize()[..COOKIE_SIZE],
+        );
+
+        cookie
+    }
+
+    fn nonce(&self) -> [u8; COOKIE_NONCE_SIZE] {
+        let mut nonce = [0u8; COOKIE_NONCE_SIZE];
+
+        let ctr0 = self.nonce_ctr0.fetch_add(1, Ordering::Relaxed);
+        let ctr1 = if ctr0 == 0 {
```

Comment:
Are you expecting u64 to wrap around? At 4Ghz that would be one wrap per 140 years.